[19.03 backport] Added garbage collector for image layers

thaJeztah opened the pull request:
~**based on top of https://github.com/docker/engine/pull/267 - only last commit is new, but marked it WIP**~

backport of

- [x] https://github.com/moby/moby/pull/39193 Added garbage collector for image layers
- [x] https://github.com/moby/moby/pull/39862 Implement code review comments in 39748
  - replaces https://github.com/moby/moby/pull/39748 Avoid using invalid value of cacheID on fail path
- [x] https://github.com/moby/moby/pull/39591 Add extra permission check in removal test
- [x] https://github.com/moby/moby/pull/39715 Unit test for getOrphan

Minor conflict because https://github.com/moby/moby/commit/072400fc4b8d6a38a2007d41072b765666a4c288 (https://github.com/moby/moby/pull/38377) is not in the 19.03 branch;

```patch
diff --cc internal/test/daemon/daemon.go
index bb8451db8c,4705d2f4ee..0000000000
--- a/internal/test/daemon/daemon.go
+++ b/internal/test/daemon/daemon.go
@@@ -60,16 -61,18 +61,31 @@@ type Daemon struct
        UseDefaultHost    bool
        UseDefaultTLSHost bool

++<<<<<<< HEAD
 +      id            string
 +      logFile       *os.File
 +      cmd           *exec.Cmd
 +      storageDriver string
 +      userlandProxy bool
 +      execRoot      string
 +      experimental  bool
 +      init          bool
 +      dockerdBinary string
 +      log           logT
++=======
+       id                         string
+       logFile                    *os.File
+       cmd                        *exec.Cmd
+       storageDriver              string
+       userlandProxy              bool
+       defaultCgroupNamespaceMode string
+       execRoot                   string
+       experimental               bool
+       init                       bool
+       dockerdBinary              string
+       log                        logT
+       imageService               *images.ImageService
++>>>>>>> 213681b66a... First step to implement full garbage collector for image layers

        // swarm related field
        swarmListenAddr string
```





**- What I did**
Added garbage collector for image layers

Closes #38488 and most probably #39247 too.
and with #38401 also https://github.com/docker/for-win/issues/745

**- How I did it**
Added logic which marks layers to be under removal by renaming layer digest folder to contain **-\<int\>-removing** suffix. That way we are able keep track that which layers are not anymore complete (=> which need to be re-downloaded if needed) and we are also able run clean up routine for those. 

**- How to verify it**
Reproduce issue like it is described on #38488
also notice bug with missing content on comment https://github.com/moby/moby/issues/38488#issuecomment-492801678

Run daemon on debug mode and stop it.

You will see log like this and orphan layer will disappear from disk.
```
DEBU[2019-05-17T02:22:21.349871543+03:00] start clean shutdown of all containers with a 15 seconds timeout... 
DEBU[2019-05-17T02:22:21.350859348+03:00] found 2 orphan layers                        
DEBU[2019-05-17T02:22:21.350918187+03:00] removing orphan layer, chain ID: sha256:3fc64803ca2de7279269048fe2b8b3c73d4536448c87c32375b2639ac168a48b , cache ID: 5ccabfe0f839144bcbb30b7eae3d65e89fd93e34b411a9ba37b753c6cadc939c 
DEBU[2019-05-17T02:22:21.352871349+03:00] Removing folder: /var/lib/docker/image/overlay2/layerdb/sha256/3fc64803ca2de7279269048fe2b8b3c73d4536448c87c32375b2639ac168a48b-0-removing 
DEBU[2019-05-17T02:22:21.353409212+03:00] Removing folder: /var/lib/docker/image/overlay2/layerdb/sha256/3fc64803ca2de7279269048fe2b8b3c73d4536448c87c32375b2639ac168a48b-1-removing 
DEBU[2019-05-17T02:22:21.353901755+03:00] removing orphan layer, chain ID: sha256:3fc64803ca2de7279269048fe2b8b3c73d4536448c87c32375b2639ac168a48b , cache ID: 4594ad5ca4f2ae9046141506d857fe9e94d68727809813d653b98353a43dd5fc 
DEBU[2019-05-17T02:22:21.356055301+03:00] Unix socket /var/run/docker/libnetwork/60b87c485747cb08641761e12ba27e169f4a9430efd4748121f1ce4d0fd3d536.sock doesn't exist. cannot accept client connections 
DEBU[2019-05-17T02:22:21.356174192+03:00] Cleaning up old mountid : start.             
INFO[2019-05-17T02:22:21.356225827+03:00] stopping event stream following graceful shutdown  error="<nil>" module=libcontainerd namespace=moby
DEBU[2019-05-17T02:22:21.356871850+03:00] Cleaning up old mountid : done.              
DEBU[2019-05-17T02:22:21.357264017+03:00] Clean shutdown succeeded 
```

and if you do same test on Windows this is what you will see:
```
time="2019-05-16T23:31:11.505667200Z" level=debug msg="start clean shutdown of all containers with a 35 seconds timeout..."
time="2019-05-16T23:31:11.510504700Z" level=debug msg="found 1 orphan layers"
time="2019-05-16T23:31:11.511193900Z" level=debug msg="removing orphan layer, chain ID: sha256:cc83dc73fcbf0e8b0ce6bce79e23643e9ffc7773af69f831e7d1f2f5764d44b2 , cache ID: bc365c16a135afa6ee813b7f550d7723d135157a0080dc1c25489314f938b656"
time="2019-05-16T23:31:11.511193900Z" level=debug msg="hcsshim::GetComputeSystems - Begin Operation"
time="2019-05-16T23:31:11.512310200Z" level=debug msg="HCS ComputeSystem Query" json="{}"
time="2019-05-16T23:31:11.513192900Z" level=debug msg="hcsshim::GetComputeSystems - End Operation - Success"
time="2019-05-16T23:31:11.514189800Z" level=debug msg="hcsshim::DestroyLayer" path="C:\\ProgramData\\docker\\windowsfilter\\bc365c16a135afa6ee813b7f550d7723d135157a0080dc1c25489314f938b656-removing"
time="2019-05-16T23:31:11.678774400Z" level=debug msg="hcsshim::DestroyLayer - succeeded" path="C:\\ProgramData\\docker\\windowsfilter\\bc365c16a135afa6ee813b7f550d7723d135157a0080dc1c25489314f938b656-removing"
time="2019-05-16T23:31:11.678774400Z" level=debug msg="Removing folder: C:\\ProgramData\\docker\\image\\windowsfilter\\layerdb\\sha256\\cc83dc73fcbf0e8b0ce6bce79e23643e9ffc7773af69f831e7d1f2f5764d44b2-0-removing"
time="2019-05-16T23:31:11.687751900Z" level=debug msg="Clean shutdown succeeded"
```

**- A picture of a cute animal (not mandatory but encouraged)**
![image](https://user-images.githubusercontent.com/6213926/57404181-09a80800-71e4-11e9-9c4a-2f870d289b8d.png)

olljanat committed:
First step to implement full garbage collector for image layers

Refactored exiting logic on way that layers are first marked to be under
removal so if actual removal fails they can be found from disk and
cleaned up.

Full garbage collector will be implemented as part of containerd
migration.

Signed-off-by: Olli Janatuinen <olli.janatuinen@gmail.com>
(cherry picked from commit 213681b66a41d7728445e30125176834ce349c2a)
Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
Added this to the 19.03.1 milestone so that we can further discuss if we want this backported for that patch release

/cc @olljanat @cpuguy83

olljanat commented:
I would like to get this backported as we see those issues on production but other hand those have been there with Windows containers from day one (logic which I needed to change was introduced by #17924 long before Windows containers was even supported) so if there is no other users reporting about this issue then it probably is not that important to get back ported.

It is also worth to note that as it took for a while me to get fix implemented we have also created workaround to it by adding some randomness to image build scripts so layers will not been re-used on newer versions (ugly way but at least it works).

crosbymichael committed:
Add extra permission check in removal test

Signed-off-by: Michael Crosby <crosbymichael@gmail.com>
(cherry picked from commit d6cbeee470105df403b7ae716c923fd11003b67c)
Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

olljanat commented:
@thaJeztah @cpuguy83 so how it is, can we carry this as part of 19.03.2 ?

ddebroy commented on layer/filestore.go:
```diff
@@ -305,6 +305,47 @@ func (fms *fileMetadataStore) GetMountParent(mount string) (ChainID, error) {
+func (fms *fileMetadataStore) getOrphan() ([]roLayer, error) {
+	var orphanLayers []roLayer
+	for _, algorithm := range supportedAlgorithms {
+		fileInfos, err := ioutil.ReadDir(filepath.Join(fms.root, string(algorithm)))
```
Are there any scenarios where the paths here may be volume GUID prefixed?

olljanat replied:
No. Those layer metadata folders are on format:
`/var/lib/docker/image/<storage driver>/layerdb/<hash algorithm>/<digest>`
example: `/var/lib/docker/image/overlay2/layerdb/sha256/ffc4c11463ee21b7532b63abd6079393c619a5d0f4b00397a4b9d1cf9efc4d9b`

and when deleteLayer() function is called it will rename it to:
`/var/lib/docker/image/overlay2/layerdb/sha256/ffc4c11463ee21b7532b63abd6079393c619a5d0f4b00397a4b9d1cf9efc4d9b-removing`
which can be then found by getOrphan()

ddebroy commented on layer/filestore.go:
```diff
@@ -305,6 +305,47 @@ func (fms *fileMetadataStore) GetMountParent(mount string) (ChainID, error) {
 	return ChainID(dgst), nil
 }
 
+func (fms *fileMetadataStore) getOrphan() ([]roLayer, error) {
```
Is it possible to have a unit-test around the logic in getOrphan in filestore_test.go please?

olljanat replied:
Sure. Implemented it on moby/moby#39715

ddebroy reviewed (commented):
I don't have know this codebase much - pointed out a couple of concerns.

olljanat committed:
Unit test for getOrphan

Signed-off-by: Olli Janatuinen <olli.janatuinen@gmail.com>
(cherry picked from commit 8660330173e5053e274cf12860079f132cbaa9fa)
Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

Commit:
Reimplement iteration over fileInfos in getOrphan.

1. Reduce complexity due to nested if blocks by using early
return/continue
2. Improve logging

Changes suggested as a part of code review comments in 39748

Signed-off-by: Vikram bir Singh <vikrambir.singh@docker.com>
(cherry picked from commit ebf12dbda08633375ab12387255d3f617ee9be38)
Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
cherry-picked https://github.com/moby/moby/pull/39715; removing "WIP"

thaJeztah commented:
ping @ddebroy @vikramhh PTAL - I think we have all parts in this PR now

vikramhh commented:
LGTM

thaJeztah commented:
@kolyshkin @andrewhsu PTAL

olljanat commented:
ping @ddebroy @kolyshkin @andrewhsu 
I found broken layer issue once again from our servers and would like to get this one forward.

ddebroy commented:
I had a couple of concerns that were addressed above. So LGTM from me (not a maintainer and don't have a lot of context around this code).

cpuguy83 commented:
Just to voice a concern emphatic I expressed in Slack only before...

I am worried about bringing this change into a mature release.

thaJeztah commented:
So, should we close this one as being too risky?

(if so, remind me to revisit https://github.com/docker/engine/pull/384 and see if I can modify that one to work without this PR)

olljanat commented:
Fair enough. Let's close this one here then and make sure that it works correctly on next main release.
Most probably this one https://github.com/moby/moby/pull/38401 (or modified version of it) will be also needed.

And FYI for anyone who might read this is that I created custom package with this one included https://github.com/olljanat/moby/releases/tag/19.03.5-olljanat1 and we will start using it on our environments.

As side comment. Based on testings today I noticed on one environment that at least Windows Defender causes time to time issues on layer handling even when I have excludes for C:\ProgramData\docker folder and docked.exe/docker.exe processes (which is one of the reasons that garbage collector is needed after all).

Drewster727 commented:
@thaJeztah any idea on a target date or version for this? I have windows nodes chewing up disk space that require regular maintenance :)

thaJeztah commented:
moved to https://github.com/moby/moby/pull/40412
